connect/api/src/main/java/org/apache/kafka/connect/data/Values.java

```diff
@@ -0,0 +1,1168 @@
+        }
+        return null;
+    }
+
+
+    /**
+     * Parse the specified string representation of a value into its schema and value.
+     *
+     * @param value the string form of the value
+     * @return the schema and value; never null, but whose schema and value may be null
+     * @see #convertToString
+     */
+    public static SchemaAndValue parseString(String value) {
+        if (value == null) {
+            return NULL_SCHEMA_AND_VALUE;
+        }
+        if (value.isEmpty()) {
+            return new SchemaAndValue(Schema.STRING_SCHEMA, value);
+        }
+        Parser parser = new Parser(value);
+        return parse(parser, false);
+    }
+
+    /**
+     * Convert the value to the desired type.
+     *
+     * @param toSchema   the schema for the desired type; may not be null
+     * @param fromSchema the schema for the supplied value; may be null if not known
+     * @return the converted value; never null
+     * @throws DataException if the value could not be converted to the desired type
+     */
+    protected static Object convertTo(Schema toSchema, Schema fromSchema, Object value) throws DataException {
+        if (value == null) {
+            if (toSchema.isOptional()) {
+                return null;
+            }
+            throw new DataException("Unable to convert a null value to a non-null value");
+        }
+        switch (toSchema.type()) {
+            case BYTES:
+                if (Decimal.LOGICAL_NAME.equals(toSchema.name())) {
+                    if (value instanceof byte[]) {
+                        return Decimal.toLogical(toSchema, (byte[]) value);
+                    }
+                    if (value instanceof BigDecimal) {
+                        return value;
+                    }
+                    if (value instanceof Number) {
+                        // Not already a decimal, so treat it as a double ...
+                        double converted = ((Number) value).doubleValue();
+                        return new BigDecimal(converted);
+                    }
+                    if (value instanceof String) {
+                        return new BigDecimal(value.toString()).doubleValue();
+                    }
+                }
+                if (value instanceof byte[]) {
+                    return value;
+                }
+                if (value instanceof BigDecimal) {
+                    return Decimal.fromLogical(toSchema, (BigDecimal) value);
+                }
+                break;
+            case STRING:
+                StringBuilder sb = new StringBuilder();
+                append(sb, value, false);
+                return sb.toString();
+            case BOOLEAN:
+                if (value instanceof Boolean) {
+                    return value;
+                }
+                if (value instanceof String) {
+                    SchemaAndValue parsed = parseString(value.toString());
+                    if (parsed.value() instanceof Boolean) {
```

Comment:
Line 405 could instead be:

    if (parsed.schema().type() == Type.BOOLEAN) {

but that didn't really seem that much better. 

The string parser always returns a SchemaAndValue, largely because the schema of any nested values needs to be known for the schema of the container (e.g., an array of boolean values vs an array of int8 values). We don't need that here because these are unwrapped primitives.

Reply:
fair enough, just felt a bit awkward. with very large chunks of code like this, it is hard to understand the reasoning without having tried to write it / deeply modify it yourself.